### PR TITLE
feat: add `endpointAdapterFactory`

### PR DESCRIPTION
**Pull request by LuisMartinez15**
EX-5388

## Motivation and context
Ability to create create `EndpointAdapter` objects with a set of `options`.
Ability to extend an existing `EndpointAdapter` with a new set of  `options`. In this case, the previous `EndpointAdapter` remains unmodified, and a new one is created merging the original options with the new ones.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that causes existing functionality to not work as expected)
- [ ] Change requires a documentation update

## How has this been tested?

Added unit tests.

## Checklist:

- [x] My code follows the **[style guidelines](./CONTRIBUTING.md#style-guides)** of this project.
- [x] I have performed a **self-review** on my own code.
- [x] I have **commented** my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the **[documentation](./CONTRIBUTING.md#documentation-style-guide)**.
- [x] My changes generate **no new warnings**.
- [x] I have added **tests** that prove my fix is effective or that my feature works.
- [x] New and existing **unit tests pass locally** with my changes.

### CODE REVIEW

**Comment by LuisMartinez15** [comment]
@javieri-empathy @diegopf
I've pushed a commit removing obsolete code that is no longer used and removing also some inline comments that, imo, were too verbose in self-explanatory code.

File: packages/x-adapter/src/__tests__/endpoint-adapter.factory.spec.ts [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+      });
+    });
+
+    // eslint-disable-next-line max-len
+    it('should combine the provided EndpointAdapterOptions.endpoint with the requestOptions.endpoint when the former is a function', async () => {
+      const requestOptions: RequestOptions = {
+        endpoint: 'https://api.empathy.co/test'
+      };
+      const endpoint: Mapper<string, string> = from => from.replace(/test/g, 'mapped');
+      const expectedEndpoint = endpoint(requestOptions.endpoint!, {});
```
Comment (javieri-empathy) [inline packages/x-adapter/src/__tests__/endpoint-adapter.factory.spec.ts:188]: I have seen this multiple times already. Why you define expected values at the top of the test? That is confusing. Usually we structure tests following the AAA pattern (I thought this pattern was called differently honestly, I had to check it):

- Arrange: Do the necessary preparations
- Act: Execute the code that we are testing
- Assert: Run the needed assertions over the result of the executed code

By defining expected values at the top of the test is like you are mixing intentions.

File: packages/x-adapter/src/endpoint-adapter.factory.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    );
+  };
+
+  endpointAdapter.extends = <NewRequest, NewResponse>(
+    extendedOptions: Partial<EndpointAdapterOptions<NewRequest, NewResponse>>
+  ) =>
+    endpointAdapterFactory<Request & NewRequest, Response & NewResponse>({
+      ...options,
+      ...extendedOptions
+    } as EndpointAdapterOptions<Request & NewRequest, Response & NewResponse>);
```
Comment (javieri-empathy) [inline packages/x-adapter/src/endpoint-adapter.factory.ts:74]: The method says `extend`, but here we are overriding. With the new mappers this method shouldn't be needed as much as it was initially thought, so I would double check if we need it, and in case we need it if it should extend (apply both mappers, the old and the new), or just override and use the last provided one.

When I did the types I thought it needed to extend, that's why the complex intersection types. If that is not needed we can simplify types.

File: packages/x-adapter/.eslintrc.js [hunk 3]
```diff
@@ -1,4 +1,5 @@
 module.exports = {
+  extends: ['plugin:@empathyco/x/recommended'],
```
Comment (javieri-empathy) [inline packages/x-adapter/.eslintrc.js:2]: Why did you add that config, it should inherit the root `.eslintrc.js` file, which already adds the recommended eslint-plugin-x config

**Review by javieri-empathy — CHANGES_REQUESTED** [review]

File: packages/x-adapter/src/endpoint-adapter.factory.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    const requestParameters =
+      requestMapper?.(request, { endpoint }) ??
+      (request as unknown as Record<string, string | boolean | number>);
+
+    /**
+     * Returns the response obtained by calling the {@link EndpointAdapterOptions.httpClient}.
+     *
+     * * If {@link EndpointAdapterOptions.httpClient} is defined:
+     * * * If {@link EndpointAdapterOptions.responseMapper} is defined, the response is mapped.
+     * * * Else, the raw response is returned.
```
[the remark this answers is not in the document]
Comment (LuisMartinez15) [reply packages/x-adapter/src/endpoint-adapter.factory.ts]: I've already removed these inline comments. We has written them while pseudocoding the logic and thought they were good for understanding the code. However, in the end, they only add verbosity and the code is self-explanatory.

File: packages/x-adapter/src/endpoint-adapter.factory.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+ * @public
+ */
+export const endpointAdapterFactory: EndpointAdapterFactory = <Request, Response>(
+  options: EndpointAdapterOptions<Request, Response>
+) => {
+  /**
+   * Local copy of the passed {@link EndpointAdapterOptions}, so they can be merged with the options
+   * passed to {@link EndpointAdapter.extends}.
+   */
+  const adapterOptions = { ...options };
```
[the remark this answers is not in the document]
Comment (LuisMartinez15) [reply packages/x-adapter/src/endpoint-adapter.factory.ts]: This was obsolete code from previous iterations. I've already removed it.

File: packages/x-adapter/.eslintrc.js (hunk 3 above)
Comment (LuisMartinez15) [reply packages/x-adapter/.eslintrc.js:2]: My bad, totally forgot it was already set up there.

File: packages/x-adapter/src/__tests__/endpoint-adapter.factory.spec.ts [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+ * @param options - The CreateEndpointAdapterFactoryOptions to create the API with.
+ *
+ * @returns The API for testing the {@link EndpointAdapter}.
+ */
+function createEndpointAdapterFactoryOptions<Request, Response>(
+  {
+    options,
+    rawResponse = {} as Response
+  }: CreateEndpointAdapterFactoryOptions<Request, Response> = {
+    rawResponse: {} as Response
```
[the remark this answers is not in the document]
Comment (LuisMartinez15) [reply packages/x-adapter/src/__tests__/endpoint-adapter.factory.spec.ts:19]: I forgot to remove this assignment after making the `rawResponse` optional in the `CreateEndpointAdapterFactoryOptions`.

File: packages/x-adapter/src/__tests__/endpoint-adapter.factory.spec.ts (hunk 1 above)
Comment (LuisMartinez15) [reply packages/x-adapter/src/__tests__/endpoint-adapter.factory.spec.ts:188]: You're totally right. Thanks for the sharing the pattern!

File: packages/x-adapter/src/types/adapter.types.ts [hunk 7]
```diff
@@ -55,7 +55,7 @@ export interface EndpointAdapterOptions<Request, Response> {
    * The endpoint to request the information to. If the endpoint can only be decided at run-time you
    * can ignore this property and use {@link RequestOptions.endpoint} parameter.
    */
-  endpoint?: string | Mapper<Request, string>;
+  endpoint?: string | Mapper<string, string>;
```
[the remark this answers is not in the document]
Comment (LuisMartinez15) [reply packages/x-adapter/src/types/adapter.types.ts:58]: I thought `Request` was a mistake, as I don't see the use case to map a request (which is usually of the form:

```
env: test
scope: desktop
device: mobile
origin: search_box:none
start: 0
rows: 24
query: shirt
lang: en
```

) to an `endpoint`. Given your input, I'll restore the previous type.

File: packages/x-adapter/src/endpoint-adapter.factory.ts [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    const [endpoint, restRequestOptions] = getOptions();
+    /**
+     * Object containing the {@link RequestOptions.parameters} used by the
+     * {@link EndpointAdapterOptions.httpClient}.
+     *
+     * If {@link EndpointAdapterOptions.requestMapper} is defined, the request used is mapped.
+     * Otherwise, we fall back to the original request.
+     */
+    const requestParameters =
+      requestMapper?.(request, { endpoint }) ??
```
[the remark this answers is not in the document]
Comment (LuisMartinez15) [reply packages/x-adapter/src/endpoint-adapter.factory.ts:55]: Changed the types from `Record<string, string | number | boolean>` to `Record<string, unknown>`

File: packages/x-adapter/src/endpoint-adapter.factory.ts [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+  options: EndpointAdapterOptions<Request, Response>
+) => {
+  /**
+   * Local copy of the passed {@link EndpointAdapterOptions}, so they can be merged with the options
+   * passed to {@link EndpointAdapter.extends}.
+   */
+  const adapterOptions = { ...options };
+
+  const endpointAdapter: EndpointAdapter<Request, Response> = (request, requestOptions) => {
+    const { endpoint: rawEndpoint, httpClient, requestMapper, responseMapper } = adapterOptions;
```
[the remark this answers is not in the document]
Comment (LuisMartinez15) [reply packages/x-adapter/src/endpoint-adapter.factory.ts]: Done

File: packages/x-adapter/src/types/adapter.types.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (LuisMartinez15) [reply packages/x-adapter/src/types/adapter.types.ts:58]: My bad. I didn't see the use case here using the `request`, but after watching it deeply I see that it should work as expected. I'll restore it back.

File: packages/x-adapter/src/endpoint-adapter.factory.ts [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+     * omitting its endpoint.
+     * @internal
+     */
+    function getOptions(): [endpoint: string, requestOptions: Omit<RequestOptions, 'endpoint'>] {
+      const { endpoint: requestOptionsEndpoint, ...restRequestOptions } = requestOptions ?? {};
+      let endpoint = requestOptionsEndpoint ?? '';
+      if (typeof rawEndpoint === 'string') {
+        endpoint = rawEndpoint;
+      } else if (typeof rawEndpoint === 'function') {
+        endpoint = rawEndpoint(endpoint, {});
```
[the remark this answers is not in the document]
Comment (LuisMartinez15) [reply packages/x-adapter/src/endpoint-adapter.factory.ts:48]: Done

File: packages/x-adapter/src/endpoint-adapter.factory.ts [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+      (request as unknown as Record<string, string | boolean | number>);
+
+    /**
+     * Returns the response obtained by calling the {@link EndpointAdapterOptions.httpClient}.
+     *
+     * * If {@link EndpointAdapterOptions.httpClient} is defined:
+     * * * If {@link EndpointAdapterOptions.responseMapper} is defined, the response is mapped.
+     * * * Else, the raw response is returned.
+     *
+     * * Else, an empty object is returned as the response.
```
[the remark this answers is not in the document]
Comment (LuisMartinez15) [reply packages/x-adapter/src/endpoint-adapter.factory.ts]: Completely missed that part. We'll leave a `TODO` with a very basic implementation.

File: packages/x-adapter/src/__tests__/endpoint-adapter.factory.spec.ts [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+    expect(mockedRequestMapper).toHaveBeenCalledTimes(1);
+    expect(mockedHttpClient).toHaveBeenCalledTimes(1);
+    expect(mockedHttpClient).toHaveBeenCalledWith(options.endpoint, {
+      parameters: request
+    });
+    expect(mockedResponseMapper).toHaveBeenCalledTimes(1);
+    expect(response).toEqual(rawResponse);
+  });
+
+  it('should extend the initial endpointAdapter options with the new ones', async () => {
```
Comment (diegopf) [inline packages/x-adapter/src/__tests__/endpoint-adapter.factory.spec.ts:73]: Maybe it is too much but I think would be nice to add some check that extending the adapter does not modify the original one.

**Review by diegopf — CHANGES_REQUESTED** [review]
Overall looks pretty good!

**Comment by tiborux** [comment]
>

File: packages/x-adapter/src/endpoint-adapter.factory.ts [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+import {
+  EndpointAdapter,
+  EndpointAdapterFactory,
+  EndpointAdapterOptions
+} from './types/adapter.types';
+import { Mapper } from './types/mapper.types';
+import { interpolate } from './utils/interpolate';
+
+/**
+ * Factory to create {@link EndpointAdapter | endpoint adapters} with the passed
```
[the remark this answers is not in the document]
Comment (LuisMartinez15) [reply packages/x-adapter/src/endpoint-adapter.factory.ts:12]: Done

File: packages/x-adapter/src/endpoint-adapter.factory.ts [hunk 14]
```diff
@@ -0,0 +1,74 @@
+import { fetchHttpClient } from './http-clients/fetch.http-client';
```
[the remark this answers is not in the document]
Comment (LuisMartinez15) [reply packages/x-adapter/src/endpoint-adapter.factory.ts:1]: I consider this to be the `entry` point of the library, so I wouldn't do that. Instead, I'd place in their respective `/¨mappers` or `/schemas` folder.

File: packages/x-adapter/src/endpoint-adapter.factory.ts [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+     *
+     * @returns The endpoint.
+     * @internal
+     */
+    function getEndpoint<Request>(
+      endpoint: string | Mapper<Request, string> | undefined,
+      request: Request
+    ): string {
+      if (!endpoint) {
+        throw Error('Tried to make a request without an endpoint');
```
[the remark this answers is not in the document]
Comment (LuisMartinez15) [reply packages/x-adapter/src/endpoint-adapter.factory.ts:49]: I don't think extracting this error message to a `const` provides any value 😅

File: packages/x-adapter/src/endpoint-adapter.factory.ts [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+        : interpolate(endpoint, request as unknown as Record<string, string | number | boolean>);
+    }
+
+    const endpoint = getEndpoint(rawEndpoint ?? requestEndpoint, request);
+    const requestParameters = requestMapper(request, { endpoint });
+
+    return httpClient<Response>(endpoint, {
+      ...requestOptions,
+      parameters: requestParameters
+    }).then(response => responseMapper(response, { endpoint, requestParameters }));
```
[the remark this answers is not in the document]
Comment (LuisMartinez15) [reply packages/x-adapter/src/endpoint-adapter.factory.ts:62]: I'd use `async/await` if after resolving the `promise` there were more code to be executed. But, in this case, we are returning the `promise` directly, so I don't think it provides any value.

**Review by diegopf — DISMISSED** [review]

File: packages/x-adapter/src/endpoint-adapter.factory.ts [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+    function getEndpoint<Request>(
+      endpoint: string | Mapper<Request, string> | undefined,
+      request: Request
+    ): string {
+      if (!endpoint) {
+        throw Error('Tried to make a request without an endpoint');
+      }
+      return typeof endpoint === 'function'
+        ? endpoint(request, {})
+        : interpolate(endpoint, request as unknown as Record<string, string | number | boolean>);
```
Comment (javieri-empathy) [inline packages/x-adapter/src/endpoint-adapter.factory.ts:53]: Modify also the `interpolate` function so it accepts `null` and `undefined` values. With this cast we are allowing `null` or `undefined` values to be passed, which will cause an error:

```ts
export function interpolate(
  string: string,
  parameters: Record<string, unknown>
): string {
```

Probably just ignoring nullish values is enough:

```ts
        parameters[property] != null ? `${head}${String(parameters[property])}${tail}` : ''
```

**Review by javieri-empathy — CHANGES_REQUESTED** [review]

File: packages/x-adapter/src/mappers/__tests__/identity.mapper.spec.ts [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+    const stringEntity = 'patata';
+    const undefinedEntity = undefined;
+
+    expect(identityMapper(arrayEntity)).toEqual(arrayEntity);
+    expect(identityMapper(booleanEntity)).toEqual(booleanEntity);
+    expect(identityMapper(nullEntity)).toEqual(nullEntity);
+    expect(identityMapper(numberEntity)).toEqual(numberEntity);
+    expect(identityMapper(objectEntity)).toEqual(objectEntity);
+    expect(identityMapper(stringEntity)).toEqual(stringEntity);
+    expect(identityMapper(undefinedEntity)).toEqual(undefinedEntity);
```
[the remark this answers is not in the document]
Comment (LuisMartinez15) [reply packages/x-adapter/src/mappers/__tests__/identity.mapper.spec.ts:19]: Done

File: packages/x-adapter/src/types/adapter.types.ts [hunk 19]
```diff
@@ -25,13 +25,14 @@ export interface EndpointAdapter<Request, Response> {
...
+   * {@link EndpointAdapter} object.
    *
-   * @param options - New options to extend the adapter with.
+   * @param options - New options to extend the {@link EndpointAdapter} with.
    */
   extends: <NewRequest, NewResponse>(
     options: Partial<EndpointAdapterOptions<NewRequest, NewResponse>>
-  ) => EndpointAdapter<Request & NewRequest, Response & NewResponse>;
+  ) => EndpointAdapter<NewRequest, NewResponse>;
 }
```
[the remark this answers is not in the document]
Comment (LuisMartinez15) [reply packages/x-adapter/src/types/adapter.types.ts:36]: Spot on. It makes total sense. Done!

File: packages/x-adapter/src/endpoint-adapter.factory.ts (hunk 17 above)
Comment (LuisMartinez15) [reply packages/x-adapter/src/endpoint-adapter.factory.ts:53]: Thanks for the refactor. Done

File: packages/x-adapter/src/endpoint-adapter.factory.ts [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+    }
```
[the remark this answers is not in the document]
Comment (LuisMartinez15) [reply packages/x-adapter/src/endpoint-adapter.factory.ts:54]: Ops. Forgot about that. Done!

**Review by javieri-empathy — APPROVED** [review]

**Review by javieri-empathy — APPROVED** [review]

**Review by diegopf — APPROVED** [review]
